runtime/kphp_tracing_binlog.cpp

```diff
@@ -0,0 +1,258 @@
+  }
+  int cur_size = get_cur_chunk_size();
+  if (cur_size + reserve_bytes > BUF_CHUNK_SIZE) {
+    finish_cur_chunk_start_next();
+  }
+}
+
+void tracing_binary_buffer::finish_cur_chunk_start_next() {
+  if (last_chunk != nullptr) {
+    last_chunk->size_bytes = get_cur_chunk_size();
+  }
+
+  char *mem = reinterpret_cast<char *>(use_heap_memory ? dl::heap_allocate(BUF_CHUNK_SIZE) : dl::allocate(BUF_CHUNK_SIZE));
+  one_chunk *chunk = reinterpret_cast<one_chunk *>(mem);
+  chunk->buf = reinterpret_cast<int *>(mem + sizeof(one_chunk));
+  chunk->size_bytes = 0;
+  chunk->prev_chunk = last_chunk;
+
+  last_chunk = chunk;
+  pos = last_chunk->buf;
+}
+
+void tracing_binary_buffer::clear(bool real_deallocate) {
+  // for script memory, no need of dl::deallocate(), because clear() is called at the end of the script,
+  // we just need to reset pointers, so that dl::allocate() is called again for a new script;
+  // but for heap memory, we need to really deallocate all chunks
+
+  if (real_deallocate) {
+    for (one_chunk *chunk = last_chunk; chunk;) {
+      one_chunk *prev_chunk = chunk->prev_chunk;
+      dl::heap_deallocate(chunk, BUF_CHUNK_SIZE);
+      chunk = prev_chunk;
+    }
+  }
+
+  last_chunk = nullptr;
+  pos = nullptr;
+}
+
+void tracing_binary_buffer::write_event_type(EventTypeEnum event_type, int custom24bits) {
+  if (unlikely(custom24bits < 0 || custom24bits >= 1 << 24)) {
+    php_warning("custom24bits %d overflow next to event_type %d", custom24bits, static_cast<int>(event_type));
+  }
+  alloc_next_chunk_if_not_enough(128);
+  write_uint32(static_cast<unsigned int>(custom24bits << 8) + static_cast<unsigned int>(event_type));
+}
+
+int tracing_binary_buffer::calc_total_size() const {
+  int binlog_size = get_cur_chunk_size();
+  for (const one_chunk *chunk = last_chunk->prev_chunk; chunk; chunk = chunk->prev_chunk) {
+    binlog_size += chunk->size_bytes;
+  }
+  return binlog_size;
+}
+
+void tracing_binary_buffer::write_int64(int64_t v) {
+  int64_t *pos64 = reinterpret_cast<int64_t *>(pos);
+  *pos64 = v;
+  pos += 2;
+}
+
+void tracing_binary_buffer::write_uint64(uint64_t v) {
+  uint64_t *pos64 = reinterpret_cast<uint64_t *>(pos);
+  *pos64 = v;
+  pos += 2;
+}
+
+void tracing_binary_buffer::write_float64(double v) {
+  int64_t *pos64 = reinterpret_cast<int64_t *>(pos);
+  *pos64 = *reinterpret_cast<int64_t *>(&v);
+  pos += 2;
+}
+
+int tracing_binary_buffer::register_string_in_table(const string &v) {
+  if (unlikely(v.size() > 255)) {
+    php_warning("too large string for register_string_in_table, len %d", v.size());
+    return 0;
+  }
+
+  int64_t hash64 = v.hash();
+  const auto it = worker_process_strings_table.find(hash64);
+  if (it != worker_process_strings_table.end()) {
+    return it->second;
+  }
+
+  int idx_in_table = worker_process_strings_table.size() + 1;
+  dl::enter_critical_section();
+  worker_process_strings_table[hash64] = idx_in_table;
+  current_php_script_registered_strings.push_back(v);
```

Comment:
I don't see a problem here, because only const strings are registered in a strings table. E.g.,
```
$span->addAttributeInt('users_count', $cnt);  // will
$span->addAttributeInt($dynamic_name, $cnt); // will not
$span->addAttributeString('upstream_backend', $cluster); // first will, second will not
```
Const strings are 
* span titles and descriptions, if const
* attribute names, if const
* @kphp-tracing codegenerated functions
Even if 10 000 functions are annotated with @kphp-tracing, plus 10 000 names+attributes (which is A LOT in fact), only 20 000 element will be stored here. Moreover, we don't store strings, we just store hashes.